activerecord/lib/active_record/associations/join_dependency/join_association.rb

```diff
@@ -93,7 +93,7 @@ def strict_loading?
           def append_constraints(connection, join, constraints)
             if join.is_a?(Arel::Nodes::StringJoin)
               join_string = Arel::Nodes::And.new(constraints.unshift join.left)
-              join.left = Arel.sql(connection.visitor.compile(join_string))
+              join.left = Arel.sql(connection.visitor.compile(join_string), retryable: true)
```

Comment:
Is `join_string` never app-supplied?

Reply:
AFAICT no, these are only constructed internally.

Reply:
FYI when `join.is_a?(Arel::Nodes::StringJoin)`, the join is a user supplied string.
e.g.
https://github.com/rails/rails/blob/284baa18c2c62abec170b61e285aefbd9604f675/activerecord/test/models/rating.rb#L7

Reply:
Ah, thanks @kamipo , I was struggling to find which codepath would test this, and settled for trying to walk through the code to deduce whether this could come from the app. I'll cut a quick PR to fix :)

Reply:
https://github.com/rails/rails/pull/51644